core/drivers/pl022_spi.c

```diff
@@ -158,155 +159,86 @@
-	PL022_SPI_MODE3 = SSPCR0_SPO1 | SSPCR0_SPH1  /* 0xC0 */
-};
-
-static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat,
-	uint8_t *rdat, size_t num_txpkts, size_t *num_rxpkts)
+static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat, uint8_t *rdat,
+	size_t num_pkts)
 {
 	size_t i = 0;
 	size_t j = 0;
 	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
 
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+	if (!num_pkts) {
+		EMSG("*num_pkts = 0!\n");
+		return;
+	}
 
-	while (i < num_txpkts) {
-		if (read8(pd->base + SSPSR) & SSPSR_TNF)
-			/* tx 1 packet */
-			write8(wdat[i++], pd->base + SSPDR);
+	if (pd->data_size_bits != 8) {
+		EMSG("data_size_bits should be 8, not %u\n",
+			pd->data_size_bits);
+		return;
 	}
 
-	do {
-		while ((read8(pd->base + SSPSR) & SSPSR_RNE) &&
-			(j < *num_rxpkts))
-			/* rx 1 packet */
-			rdat[j++] = read8(pd->base + SSPDR);
-	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
+	if (wdat) {
+		while (i < num_pkts) {
+			if (read8(pd->base + SSPSR) & SSPSR_TNF)
+				/* tx 1 packet */
+				write8(wdat[i++], pd->base + SSPDR);
+		}
+	}
 
-	*num_rxpkts = j;
+	if (rdat) {
+		while ((j < num_pkts) &&
+			(read8(pd->base + SSPSR) & SSPSR_BSY)) {
+			if (read8(pd->base + SSPSR) & SSPSR_RNE)
+				/* rx 1 packet */
+				rdat[j++] = read8(pd->base + SSPDR);
+		}
 
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+		if (j < num_pkts) {
+			EMSG("# packets requested: %zu\n", num_pkts);
+			EMSG("# packets received: %zu\n", j);
+		}
+	}
 }
 
-static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat,
-	uint16_t *rdat, size_t num_txpkts, size_t *num_rxpkts)
+static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat, uint16_t *rdat,
+	size_t num_pkts)
 {
 	size_t i = 0;
 	size_t j = 0;
 	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
 
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
-
-	while (i < num_txpkts) {
-		if (read8(pd->base + SSPSR) & SSPSR_TNF)
-			/* tx 1 packet */
-			write16(wdat[i++], pd->base + SSPDR);
+	if (!num_pkts) {
+		EMSG("*num_pkts = 0!\n");
+		return;
 	}
 
-	do {
-		while ((read8(pd->base + SSPSR) & SSPSR_RNE)
-			&& (j < *num_rxpkts))
-			/* rx 1 packet */
-			rdat[j++] = read16(pd->base + SSPDR);
-	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
-
-	*num_rxpkts = j;
-
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
-}
-
-static void pl022_tx8(struct spi_chip *chip, uint8_t *wdat,
-	size_t num_txpkts)
-{
-	size_t i = 0;
-	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
-
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
-
-	while (i < num_txpkts) {
-		if (read8(pd->base + SSPSR) & SSPSR_TNF)
-			/* tx 1 packet */
-			write8(wdat[i++], pd->base + SSPDR);
+	if (pd->data_size_bits != 16) {
+		EMSG("data_size_bits should be 16, not %u\n",
+			pd->data_size_bits);
+		return;
```

Comment:
is this an real error case ?
if so, there should be some error status propagation.

same for pl022_`tx`rx8().